### PR TITLE
Eager checkpoint

### PR DESCRIPTION
**Pull request by junjun315**
1. add checkpoint.py for eager
2. use numpy.save/load(bad)

### CODE REVIEW

File: python/paddle/fluid/imperative/checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    if all variables were saved in a single file, use `filename` to specify
+    the file name.
+
+    Args:
+        executor(Executor): The executor to run for loading persistable variables.
+        dirname(str): The directory path.
+        filename(str|None): The file which saved all variables. If variables were
+                            saved in differnet files, set it to None.
+                            Default: None
+        layer(Layer|PyLayer|None): The layer whose parameters will
```
Comment (panyx0718) [inline python/paddle/fluid/imperative/checkpoint.py]: no PyLayer

File: python/paddle/fluid/imperative/checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+# limitations under the License.
+
+from __future__ import print_function
+
+import os
+import numpy
+from .layers import Layer
+
+
+def save_persistables(dirname, filename=None, layer=None):
```
Comment (panyx0718) [inline python/paddle/fluid/imperative/checkpoint.py]: @velconia how to save persistables from optimizer?

File: python/paddle/fluid/imperative/checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        # todo lujun: if remove np.save/load, fix this
+        parameter_dict = numpy.load(os.path.join(dirname, filename))
+    else:
+        for pwd, dirs, files in os.walk(dirname):
+            for parameter_file in files:
+                # todo lujun: if remove np.save/load, fix this
+                if os.path.splitext(parameter_file)[-1] == '.npy':
+                    parameter_dict[parameter_file[:-4]] = numpy.load(
+                        os.path.join(pwd, parameter_file))
+    return parameter_dict
```
Comment (panyx0718) [inline python/paddle/fluid/imperative/checkpoint.py]: how to use this dict?

File: python/paddle/fluid/imperative/checkpoint.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+def _save_var_to_file(var_dict, file_dir, file_name):
+    # ToDo change to C++ do
+    if file_name is not None:
+        # save to np.savez
+        numpy.savez(os.path.join(file_dir, file_name), **var_dict)
+    else:
+        if not os.path.exists(file_dir):
+            os.makedirs(file_dir)
+        for var_name in var_dict.keys():
+            numpy.save(os.path.join(file_dir, var_name), var_dict[var_name])
```
Comment (panyx0718) [inline python/paddle/fluid/imperative/checkpoint.py]: why use numpy? why not use paddle's save op? this format is not compatible with static graph?

**Review by panyx0718 — COMMENTED** [review]
test?

File: python/paddle/fluid/imperative/checkpoint.py (hunk 4 above)
Comment (junjun315) [reply python/paddle/fluid/imperative/checkpoint.py]: change to paddle's op

File: python/paddle/fluid/imperative/checkpoint.py (hunk 3 above)
Comment (junjun315) [reply python/paddle/fluid/imperative/checkpoint.py]: it's a Variable dict. like : `param_dict['PtbModel_0.w_1']`. The key of the dictionary is the parameter name when it is saved( no full_name). as the file name :
`PtbModel_0.w_0
PtbModel_0.w_1
embedding_para`

File: python/paddle/fluid/imperative/checkpoint.py (hunk 1 above)
Comment (junjun315) [reply python/paddle/fluid/imperative/checkpoint.py]: ok, fixed

File: python/paddle/fluid/imperative/checkpoint.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+import os
+import contextlib
+import paddle.fluid as fluid
+from paddle.fluid import core
+from paddle.fluid.framework import Variable
+from paddle.fluid.executor import Executor
+
+
+def save_persistables(var_list, dirname, filename=None, place=None):
```
Comment (velconia) [inline python/paddle/fluid/imperative/checkpoint.py]: Need a default place and a default filename for this interface

File: python/paddle/fluid/imperative/checkpoint.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+                                                        init_cell)
+            param_path = "./my_paddle_model"
+            fluid.imperative.checkpoint.save_persistables(ptb_model.parameters, dirname=param_path,
+                                       layer=ptb_model, place=fluid.CPUPlace())
+    """
+    if var_list:
+        _save_var_to_file(var_list, place, dirname, filename)
+
+
+def load_persistables(var_list, dirname, filename=None, place=None):
```
Comment (velconia) [inline python/paddle/fluid/imperative/checkpoint.py]: same here

File: python/paddle/fluid/imperative/checkpoint.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+        """
+    if var_list:
+        return _load_var_from_file(var_list, place, dirname, filename)
+
+    return {}
+
+
+def _save_var_to_file(var_list, place, file_dir, file_name):
+    with guard(place) as train_pro:
```
Comment (velconia) [inline python/paddle/fluid/imperative/checkpoint.py]: train_prog

File: python/paddle/fluid/imperative/checkpoint.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+            init_cell_data = np.zeros(
+                (num_layers, batch_size, hidden_size), dtype='float32')
+            x = to_variable(x_data)
+            y = to_variable(y_data)
+            init_hidden = to_variable(init_hidden_data)
+            init_cell = to_variable(init_cell_data)
+            dy_loss, last_hidden, last_cell = ptb_model(x, y, init_hidden,
+                                                        init_cell)
+            param_path = "./my_paddle_model"
+            fluid.imperative.checkpoint.save_persistables(ptb_model.parameters, dirname=param_path,
```
Comment (velconia) [inline python/paddle/fluid/imperative/checkpoint.py]: Need a ut for this

File: python/paddle/fluid/imperative/checkpoint.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+from __future__ import print_function
+from paddle.fluid.wrapped_decorator import signature_safe_contextmanager
+
+import os
+import contextlib
+import paddle.fluid as fluid
+from paddle.fluid import core
+from paddle.fluid.framework import Variable
+from paddle.fluid.executor import Executor
+
```
Comment (velconia) [inline python/paddle/fluid/imperative/checkpoint.py:22]: Add ```__all__``` here, what is the interface api exported to user?

File: python/paddle/fluid/imperative/checkpoint.py (hunk 2 above)
Comment (velconia) [reply python/paddle/fluid/imperative/checkpoint.py]: optimizer need to save parameters list, we didn't save it now

File: python/paddle/fluid/imperative/checkpoint.py (hunk 5 above)
Comment (junjun315) [reply python/paddle/fluid/imperative/checkpoint.py]: fixed, remove parm about `place`

File: python/paddle/fluid/imperative/checkpoint.py (hunk 6 above)
Comment (junjun315) [reply python/paddle/fluid/imperative/checkpoint.py]: the same

File: python/paddle/fluid/imperative/checkpoint.py (hunk 7 above)
Comment (junjun315) [reply python/paddle/fluid/imperative/checkpoint.py]: ok, removed it

File: python/paddle/fluid/imperative/checkpoint.py (hunk 9 above)
Comment (junjun315) [reply python/paddle/fluid/imperative/checkpoint.py:22]: added

File: python/paddle/fluid/tests/unittests/test_imperative_checkpoint.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+                    sgd.minimize(avg_loss)
+                    mnist.clear_gradients()
+
+                    fluid.imperative.save_persistables(mnist.parameters(),
+                                                       "save_dir")
+
+                    for param in mnist.parameters():
+                        dy_param_init_value[param.name] = param._numpy()
+
+            poad_dict = fluid.imperative.load_persistables(para_list,
```
Comment (velconia) [inline python/paddle/fluid/tests/unittests/test_imperative_checkpoint.py]: How about bypass ```mnist``` here?

File: python/paddle/fluid/tests/unittests/test_imperative_checkpoint.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+                    dy_out = avg_loss._numpy()
+
+                    if epoch == 0 and batch_id == 0:
+                        para_list = [p for p in mnist.parameters()]
+
+                    avg_loss._backward()
+                    sgd.minimize(avg_loss)
+                    mnist.clear_gradients()
+
+                    fluid.imperative.save_persistables(mnist.parameters(),
```
Comment (velconia) [inline python/paddle/fluid/tests/unittests/test_imperative_checkpoint.py]: I guess should save **and load** each time and compare with the **no save and load each time**

File: python/paddle/fluid/imperative/checkpoint.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+            type='save_combine',
+            inputs={'X': save_var_list},
+            outputs={},
+            attrs={'file_path': os.path.join(file_dir, file_name)})
+
+
+def _load_var_from_file(var_list, file_dir, file_name):
+    load_block = fluid.default_main_program().global_block()
+    load_var_map = {}
+    load_var_list_fetch = []
```
Comment (velconia) [inline python/paddle/fluid/imperative/checkpoint.py]: no use?

File: python/paddle/fluid/imperative/checkpoint.py [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+
+def _clone_var_in_block_(block, var):
+    assert isinstance(var, Variable)
+    return block.create_var(
+        name=var.name,
+        shape=var.shape,
+        dtype=var.dtype,
+        type=var.type,
+        lod_level=var.lod_level,
+        persistable=True)
```
Comment (velconia) [inline python/paddle/fluid/imperative/checkpoint.py:194]: Is persistable always True, or we should check the var?

File: python/paddle/fluid/imperative/checkpoint.py [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+    load_var_map = {}
+    load_var_list_fetch = []
+
+    for each_var in var_list:
+        assert isinstance(each_var, Variable)
+        if each_var.type == core.VarDesc.VarType.RAW:
+            continue
+        new_var = _clone_var_in_block_(load_block, each_var)
+        load_var_list_fetch.append(new_var.name)
+        if file_name is None:
```
Comment (velconia) [inline python/paddle/fluid/imperative/checkpoint.py:161]: What will happen if file_name is NOT None

File: python/paddle/fluid/imperative/checkpoint.py (hunk 12 above)
Comment (junjun315) [reply python/paddle/fluid/imperative/checkpoint.py]: removed

File: python/paddle/fluid/imperative/checkpoint.py (hunk 14 above)
Comment (junjun315) [reply python/paddle/fluid/imperative/checkpoint.py:161]: save in one file

File: python/paddle/fluid/imperative/checkpoint.py (hunk 13 above)
Comment (junjun315) [reply python/paddle/fluid/imperative/checkpoint.py:194]: remove it

File: python/paddle/fluid/tests/unittests/test_imperative_checkpoint.py (hunk 11 above)
Comment (junjun315) [reply python/paddle/fluid/tests/unittests/test_imperative_checkpoint.py]: changed

File: python/paddle/fluid/tests/unittests/test_imperative_checkpoint.py (hunk 10 above)
Comment (junjun315) [reply python/paddle/fluid/tests/unittests/test_imperative_checkpoint.py]: fit it

File: python/paddle/fluid/imperative/checkpoint.py [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+    give `layer` and then trys to load these variables from the folder
+    `dirname` or the file `filename`.
+
+    Use the `dirname` to specify the folder where persistable variables were
+    saved. If variables were saved in separate files, set `filename` None;
+    if all variables were saved in a single file, use `filename` to specify
+    the file name.
+
+    Args:
+        var_list(dict of Parameters|Layer): The parameters will
```
Comment (panyx0718) [inline python/paddle/fluid/imperative/checkpoint.py:38]: why not list of parameters, why layer?

**Pull request closed without merging**